pkg/types/chain_reader.go

```diff
@@ -51,21 +68,212 @@ type BoundContract struct {
+
+func (o SortByBlock) GetDirection() SortDirection {
+	return o.dir
+}
+
+type SortBySequence struct {
+	dir SortDirection
+}
+
+func NewSortBySequence(sortDir SortDirection) SortBySequence {
+	return SortBySequence{dir: sortDir}
+}
+
+func (o SortBySequence) GetDirection() SortDirection {
+	return o.dir
+}
+
+type QueryFilter interface {
+	Accept(visitor Visitor)
+}
+
+// Where eg. usage:
+//
+//	Where(
+//		NewAndFilter(
+//			NewOrFilter(
+//				NewBlockFilter(startBlock, Gte),
+//				NewBlockFilter(endBlock, Lte),
+//				),
+//			NewOrFilter(
+//				NewTimeStampFilter(startTs, Gte),
+//				NewTimeStampFilter(endTs, Lte),
+//			)
+//		)
+//	)
+func Where(filters ...QueryFilter) []QueryFilter {
+	return append([]QueryFilter{}, filters...)
+}
+
+// TODO add ORFilter
+type AndFilter struct {
+	Filters []QueryFilter
+}
+
+func NewAndFilter(filters ...QueryFilter) *AndFilter {
+	return &AndFilter{Filters: filters}
+}
+
+func (f *AndFilter) Accept(visitor Visitor) {
+	visitor.VisitAndFilter(*f)
+}
+
+type AddressFilter struct {
+	Addresses []string
+}
+
+func NewAddressesFilter(addresses ...string) *AddressFilter {
+	return &AddressFilter{Addresses: addresses}
+}
+
+func (f *AddressFilter) Accept(visitor Visitor) {
+	visitor.VisitAddressFilter(*f)
+}
+
+type Confirmations int32
+
+const (
+	Finalized   = Confirmations(0)
+	Unconfirmed = Confirmations(1)
+)
+
+type ConfirmationsFilter struct {
+	Confirmations
+}
+
+func NewConfirmationsFilter(confs Confirmations) *ConfirmationsFilter {
+	return &ConfirmationsFilter{Confirmations: confs}
+}
+
+func (f *ConfirmationsFilter) Accept(visitor Visitor) {
+	visitor.VisitConfirmationFilter(*f)
+}
+
+type BlockFilter struct {
+	Block    uint64
+	Operator ComparisonOperator
+}
+
+func NewBlockFilter(block uint64, operator ComparisonOperator) *BlockFilter {
+	return &BlockFilter{block, operator}
+}
+
+func NewBlockRangeFilter(start, end uint64) *AndFilter {
+	return NewAndFilter(
+		NewBlockFilter(start, Gte),
+		NewBlockFilter(end, Lte),
+	)
+}
+
+func (f *BlockFilter) Accept(visitor Visitor) {
+	visitor.VisitBlockFilter(*f)
+}
+
+func NewTimestampFilter(timestamp uint64, operator ComparisonOperator) *TimestampFilter {
+	return &TimestampFilter{timestamp, operator}
+}
+
+type TimestampFilter struct {
+	Timestamp uint64
+	Operator  ComparisonOperator
+}
+
+func (f *TimestampFilter) Accept(visitor Visitor) {
+	visitor.VisitTimestampFilter(*f)
+}
+
+func NewTxHashFilter(txHash string) *TxHashFilter {
+	return &TxHashFilter{txHash}
+}
+
+type TxHashFilter struct {
+	TxHash string
+}
+
+func (f *TxHashFilter) Accept(visitor Visitor) {
+	visitor.VisitTxHashFilter(*f)
 }
 
-type EventFilter struct {
-	AddressList []string   // contract address
-	KeysList    [][]string // 2D list of indexed search keys, outer dim = AND, inner dim = OR.  Params[0] is the name of the event (or "event type"), rest are any narrowing parameters that may help further specify the event
-	Retention   time.Duration
+type Visitor interface {
```

Comment:
Is the expectation that a `Visitor` is some arbitrary filter implementation? Does this introduce chain-specific code on the product side?

Reply:
Visitor should have a chain specific implementation in the chain specific loop, but this won't be exposed to the product